subsys/net/lib/http_server/src/server.c

```diff
@@ -0,0 +1,538 @@
+static char buffer[BUFFER_SIZE] = {0};
+static char http_response[BUFFER_SIZE];
+const char *preface = "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n";
+
+struct http_parser_settings parserSettings;
+struct http_parser parser;
+
+static unsigned char settings_frame[9] = {
+    0x00, 0x00, 0x00, /* Length(24) */
+    0x04, /* Type(8) - 0x04 means setting frames used for configuration or
+	     acknowledgement */
+    0x00, /* Flags(8) - 0x00 means unused flags and 0x01 means ACK */
+    0x00, 0x00, 0x00,
+    0x00}; /* Reserved(1), Stream Identifier(31) - 0x00 means that this frame is
+	related to the overall connection, and not for a specific stream */
+
+static unsigned char settings_ack[9] = {
+    0x00, 0x00, 0x00, /* Length(24) */
+    0x04, /* Type(8) - 0x04 means setting frames used for configuration or
+	     acknowledgement */
+    0x01, /* Flags(8) - 0x01 means ACK */
+    0x00, 0x00, 0x00, 0x00}; /* Reserved(1), Stream Identifier(31) */
+
+Http2Frame frames[MAX_FRAMES];
+
+static const char content[] = {
+#include "index.html.gz.inc"
+};
+
+static const char content2[] = {
+#include "index.html.inc"
+};
+
+int hasUpgradeHeader = 1;
+
+int on_header_field(struct http_parser *parser, const char *at, size_t length)
+{
+    if (length == 7 && strncasecmp(at, "Upgrade", length) == 0) {
+	printf("The \"Upgrade: h2c\" header is present.\n");
+	hasUpgradeHeader = 0;
+    }
+    return 0;
+}
+
+int on_url(struct http_parser *parser, const char *at, size_t length)
+{
+    strncpy(url_buffer, at, length);
+    url_buffer[length] = '\0';
+    printf("Requested URL: %s\n", url_buffer);
+    return 0;
+}
+
+ssize_t sendall(int sock, const void *buf, size_t len)
+{
+    while (len) {
+	ssize_t out_len = send(sock, buf, len, 0);
+	if (out_len < 0) {
+	    return out_len;
+	}
+	buf = (const char *)buf + out_len;
+	len -= out_len;
+    }
+
+    return 0;
+}
+
+void close_client_connection(struct http2_server_ctx *ctx, int clientIndex)
+{
+    close(ctx->client_fds[clientIndex].fd);
+    ctx->client_fds[clientIndex].fd = 0;
+    ctx->client_fds[clientIndex].events = 0;
+    ctx->client_fds[clientIndex].revents = 0;
+
+    if (clientIndex == ctx->num_clients) {
+	while (ctx->num_clients > 0 &&
+	       ctx->client_fds[ctx->num_clients].fd == 0) {
+	    --ctx->num_clients;
+	}
+    }
+}
+
+int accept_new_client(int server_fd)
+{
+    int new_socket;
+    struct sockaddr_in address;
+    int addrlen = sizeof(address);
+
+    new_socket =
+	accept(server_fd, (struct sockaddr *)&address, (socklen_t *)&addrlen);
+    if (new_socket < 0) {
+	perror("accept failed");
+	return -1;
+    }
+
+    return new_socket;
+}
+
+void handle_http1_request(struct http2_server_ctx *ctx, int i)
+{
+    const char *data;
+    size_t len;
+    data = content;
+    len = sizeof(content);
+
+    http_parser_init(&parser, HTTP_REQUEST);
+    http_parser_settings_init(&parserSettings);
+    parserSettings.on_url = on_url;
+
+    http_parser_execute(&parser, &parserSettings, buffer, BUFFER_SIZE);
```

Comment:
We should provide the actual number of bytes received to the parser, not `BUFFER_SIZE`. Also, as mentioned earlier we should not assume that we've received the full request in a single receive call.

See how the HTTP client lib deals with the parser - https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/http/http_client.c#L452, it's always a recv()/http_parser_execute() pair.